### PR TITLE
Reduce code duplication in eval.rs and eval_const.rs

### PR DESCRIPTION
**Pull request by ysthakur**
# Description

This PR is for sharing code between `eval.rs` (normal runtime evaluation) and `eval_const.rs` (const evaluation). The two have a lot of the same logic, and when fixing a bug or something, it's easy to modify one and forget to modify the other. For example, when constructing records, `eval.rs` would check for duplicate columns, but `eval_const.rs` wouldn't, even though the behavior of both should have matched (#11144 made `eval_const.rs` check for duplicate columns too).

I figured the easiest way would be to make an `Eval` trait in nu-protocol, with two impls, one for runtime evaluation and one for const evaluation. The trait has an `eval` method acting like `eval_expression`/`eval_constant`. It has a huge match block for the expression being evaluated, and for cases that work differently between `eval.rs` and `eval_const.rs`, it dispatches to abstract methods (not sure what the Rust terminology here is).

# User-Facing Changes

<s>If there are any, something's gone wrong.</s>

Const eval has been modified (very slightly) to fit regular eval's behavior more closely:
- Const eval used to only allow strings as column names in records and tables. Now it also allows numbers, datetimes, etc., like regular eval. This makes Nushell more permissive, so hopefully it won't break anything.
- Error messages when const evaluating `FullCellPath` and `ValueWithUnit` now match regular eval (barely noticeable)
  - Errors encountered when evaluating the inner expression in `FullCellPath` are left as is, without converting to a generic error
  - For `ValueWithUnit`, if the value isn't an int, you now get a `CantConvert` rather than `NotAConstant`

# Tests + Formatting

Are tests necessary?

# After Submitting

### CODE REVIEW

**Comment by ysthakur** [comment]
There's a few kinds of expressions, such as `Garbage` and `Operator`, that will never be evaluated directly in practice. `eval.rs` makes these result in `Value::Nothing`, but in `eval_const.rs`, the last arm of the match is a wildcard that makes all of them result in a `NotAConstant` error. Would it be okay to make these expressions evaluate to `Value::Nothing` for const evaluation too? It shouldn't be a change in behavior, since they should never actually be evaluated, we just need to satisfy the compiler.

File: crates/nu-protocol/src/eval_base.rs [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+                // a constant, it's that it isn't an Int
+                match Self::eval(state, mut_state, e)? {
+                    Value::Int { val, .. } => unit.item.to_value(val, unit.span),
+                    x => Err(ShellError::CantConvert {
+                        to_type: "unit value".into(),
+                        from_type: x.get_type().to_string(),
+                        span: e.span,
+                        help: None,
+                    }),
+                }
```
Comment (ysthakur) [inline crates/nu-protocol/src/eval_base.rs:151]: Just commenting to draw attention to this

**Comment by ysthakur** [comment]
Currently, const eval doesn't handle `Expr::Directory` and `Expr::GlobPattern`, and it doesn't expand `Expr::Filepath`s. Could `current_dir_str_const` be moved to `nu-protocol`, so that it could handle those? (would it make sense for const eval to get the current working directory?)

**Comment by sholderbach** [comment]
> Currently, const eval doesn't handle `Expr::Directory` and `Expr::GlobPattern`, and it doesn't expand `Expr::Filepath`s. Could `current_dir_str_const` be moved to `nu-protocol`, so that it could handle those? (would it make sense for const eval to get the current working directory?)

As the working directory can change during execution anything depending in spirit on the `$env.PWD` variable can not be const under the current semantics. @kubouch may be the one to ask under which circumstances we could relax those rules.

**Comment by sholderbach** [comment]
> * There's no way this should be a change seen by users, but I'll put it here anyway: expressions such as `Expr::Garbage` that should never be evaluated now return `Value::Nothing` in const eval rather than giving a `NotAConstant` error

As const eval runs at parse time my cautious thinking (without being familiar with the implementation) is that it may be more likely that invalid expressions get passed to const evaluation e.g. during interactive parsing for highlight/completion/IDE stuff. I would guess that an explicit error would be better than something that appears normal like `Value::Nothing`.

**Comment by ysthakur** [comment]
> As the working directory can change during execution anything depending in spirit on the `$env.PWD` variable can not be const under the current semantics.

Makes sense, I'll leave it as is then.

> As const eval runs at parse time my cautious thinking (without being familiar with the implementation) is that it may be more likely that invalid expressions get passed to const evaluation e.g. during interactive parsing for highlight/completion/IDE stuff. I would guess that an explicit error would be better than something that appears normal like `Value::Nothing`.

Got it, I'll make it go back to the old behavior then.

**Comment by kubouch** [comment]
Yeah, general const eval shouldn't expand file paths. There are some cases like `use` that need to expand at parse time, but these are special parser keywords.

Also agree with sholderbach, maybe it's better to be more on the cautious side and throw an error rather than nothing on unreachable expressions.

**Comment by kubouch** [comment]
OK, to me this looks good. I think tests are not required since there are already tests covering const eval and these would break if it wasn't correct. But if you want to add tests covering some edge cases, that's always welcome.
